tensorflow/python/distribute/cluster_resolver/mpi_cluster_resolver.py

```diff
@@ -0,0 +1,97 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Implementation of a Cluster Resolver for MPI jobs."""
+
+from socket import gethostname
+from slurm_cluster_resolver import SlurmClusterResolver
+
+from tensorflow.python.util.tf_export import tf_export
+
+
+@tf_export('distribute.cluster_resolver.MPIClusterResolver')
+class MPIClusterResolver(SlurmClusterResolver):
+  """ClusterResolver for programs run via MPI (mpirun, srun, ...)
+
+  This is an implementation of ClusterResolver for MPI programs which can
+  be used for distributed TensorFlow.
+  When no explicit values are set it will retrieve all information from MPI.
+  For rank and number of Tasks the values are gotten from the MPI_COMM_WORLD
+  communicator. It also automatically resolves hostnames which requires some
+  communication on startup.
+  """
+
+  def __init__(self,
+               jobs=None,
+               port_base=8888,
+               gpus_per_node=None,
+               gpus_per_task=None,
+               auto_set_gpu=True,
+               rpc_layer='grpc'):
+    """Creates a new MPIClusterResolver object.
+
+    For any parameter not set it will query MPI_COMM_WORLD for the value.
+    With the number of GPUs per node and per task it allocates GPUs to tasks by
+    setting environment variables.
+    Using the resolver works best (and is easier) with homogeneous tasks but
+    heterogeneous tasks (number of tasks varying per node) are also possible as
+    long as the number of GPUs per task stays constant.
+
+    Args:
+      jobs: Dictionary with job names as key and number of tasks in the job as
+        value. Defaults to as many 'worker's as there are MPI tasks.
+      port_base: The first port number to start with for processes on a node.
+      gpus_per_node: Number of GPUs available on each node. Defaults to the
+        number of GPUs reported by nvidia-smi
+      gpus_per_task: Number of GPUs to be used for each task. Default is to
+        evenly distribute the gpus_per_node to tasks_per_node.
+      auto_set_gpu: Set the visible CUDA devices automatically while resolving
+        the cluster by setting CUDA_VISIBLE_DEVICES environment variable.
+        Defaults to True.
+      rpc_layer: The protocol TensorFlow used to communicate between nodes.
+        Defaults to 'grpc'.
+
+    Returns:
+      A MPIResolver object which can be used with distributed TensorFlow.
+
+    Raises:
+      RuntimeError: If requested more GPUs per node then available or
+        requested more tasks then assigned tasks.
+    """
+    from mpi4py import MPI # pylint: disable=import-outside-toplevel
```

Comment:
Maybe catch the ImportError and return a better error message directing users to install mpi4py? Something like https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/distribute/cluster_resolver/tpu_cluster_resolver.py#L32-L34 works.

Reply:
would a `raise ImportError("mpi4py could not be imported. Run 'pip install mpi4py' to install")` work for you? Logging isn't enough in this case.

Reply:
Yah works for me!

Reply:
Done in 6a47a283b0

Reply:
Sorry hat a formatting issue and had to force-push the last commit. Strange though